src/module/service.cpp

```diff
@@ -0,0 +1,290 @@
+    if (m->services.find(*subtype) == m->services.end()) {
+        return grpc::Status(grpc::UNKNOWN, "no grpc service for " + subtype->to_string());
+    }
+    std::shared_ptr<SubtypeService> svc = m->services.at(*name.to_subtype());
+    std::shared_ptr<ResourceBase> res = svc->resource(name.name);
+    if (res == nullptr) {
+        return grpc::Status(
+            grpc::UNKNOWN,
+            "unable to remove resource " + name.to_string() + " as it doesn't exist.");
+    }
+
+    try {
+        res->stop();
+    } catch (std::string err) {
+        BOOST_LOG_TRIVIAL(error) << "unable to stop resource: " << err;
+    }
+
+    if (name.resource_type == COMPONENT && !(*name.to_subtype() == Generic::subtype())) {
+        if (module->services.find(Generic::subtype()) == module->services.end()) {
+            return grpc::Status(grpc::UNKNOWN, "no generic service");
+        }
+        std::shared_ptr<SubtypeService> generic_service = module->services.at(Generic::subtype());
+        generic_service->remove(name);
+    }
+    svc->remove(name);
+    return grpc::Status();
+};
+
+::grpc::Status ModuleService_::Ready(::grpc::ServerContext* context,
+                                     const ::viam::module::v1::ReadyRequest* request,
+                                     ::viam::module::v1::ReadyResponse* response) {
+    module->lock.lock();
+    viam::module::v1::HandlerMap hm = this->module->handles.to_proto();
+    *response->mutable_handlermap() = hm;
+    parent_addr = request->parent_address();
+    module->lock.unlock();
+    response->set_ready(module->ready);
+    return grpc::Status();
+};
+
+ModuleService_::ModuleService_(std::string addr) {
+    module = std::make_shared<Module>(addr);
+}
+
+void ModuleService_::start() {
+    module->lock.lock();
+    mode_t old_mask = umask(0077);
+    int sockfd = socket(AF_UNIX, SOCK_STREAM, 0);
+    listen(sockfd, 10);
+    umask(old_mask);
+
+    // TODO(RSDK-1742) see if we can/want to do this in an init instead
+    Server::register_service(this);
+    std::string address = "unix://" + module->addr;
+    Server::add_listening_port(address);
+
+    module->lock.unlock();
+    module->set_ready();
+}
+
+ModuleService_::~ModuleService_() {
+    this->close();
+}
+
+void ModuleService_::close() {
+    // CR erodkin: this actually doesn't shutdown gracefully at all! fix that
+    BOOST_LOG_TRIVIAL(info) << "Shutting down gracefully.";
+
+    if (parent != nullptr) {
+        try {
+            parent->close();
+        } catch (std::exception& exc) {
+            BOOST_LOG_TRIVIAL(error) << exc.what();
+        }
+    }
+    Server::shutdown();
+}
+
+void ModuleService_::add_api_from_registry(Subtype api) {
+    if (module->services.find(api) != module->services.end()) {
+        return;
+    }
+    module->lock.lock();
+    std::shared_ptr<SubtypeService> new_svc = SubtypeService::of_subtype(api.resource_subtype);
+
+    std::shared_ptr<ResourceSubtype> rs = Registry::lookup_subtype(api);
```

Comment:
Note to reviewers: this isn't being used here (whoops!) but is being used in a subsequent ticket, so I'm leaving it here for now.